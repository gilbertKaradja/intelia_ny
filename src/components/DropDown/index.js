import React, { Component } from 'react';
import PropTypes from 'prop-types';

import icon from './caret-down.svg';

class DropDown extends Component {
    constructor() {
        super();

        this.state = {
            showList: false
        }

        this.openHandler = this.openHandler.bind(this);
        this.closeHandler = this.closeHandler.bind(this);
    }

    componentDidMount() {
        // document.addEventListener('click', this.closeHandler);
    }

    componentWillUnmount() {
        // document.removeEventListener('click', this.closeHandler);
    }

    render() {
        let {value, options} = this.props;
        return (
            <div className="DropDown">

                {this.state.showList && (
                    <div className="overlay" onClick={this.closeHandler}></div>
                )}

                <div className="value_container" onClick={this.openHandler}>

                    <div className="text">
                        {value.label}
                    </div>

                    <div className="icon">
                        <img src={icon} alt='caret'/>
                    </div>

                </div>

                {this.state.showList && (
                    <div className="dropdown_list">
                        <ul>
                            {options.map(item => {
                                return (
                                    <li
                                        key={item.value}
                                        className={value.value === item.value ? 'active' : ''}
                                        onClick={() => { this.itemSelectHandler(item) }}>
                                        {item.label}
                                    </li>
                                );
                            })}
                        </ul>
                    </div>
                )}
            </div>
        );
    }

    openHandler(evt) {
        evt.stopPropagation();

        let { showList } = this.state;
        if (showList) {
            return;
        }

        this.setState({ showList: true });
    }

    itemSelectHandler(value) {
        this.props.onChange(value);

        this.setState({ showList: false });
    }

    closeHandler(evt) {
        this.setState({ showList: false });
    }


}

DropDown.propTypes = {
    value: PropTypes.object,
    options: PropTypes.array,
    onChange: PropTypes.func
}

export default DropDown;


