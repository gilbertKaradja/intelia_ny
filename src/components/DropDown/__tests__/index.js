import React from 'react';
import { shallow, mount } from 'enzyme';

import DropDown from '../index.js';

let sampleOptions = [
    { label: 'Train', value: '1' },
    { label: 'Stop Sign', value: '9' },
    { label: 'Couch', value: 'joke' },
    { label: 'Random Sentence', value: 'evo993' },
    { label: 'Something', value: 'chihuahua' }
];

describe('DropDown component', () => {

    it('should render without issues', () => {

        //renders initial dropdown properly
        const component = shallow(
            <DropDown
                value={sampleOptions[0]}
                options={sampleOptions}
            />
        );
        expect(component).toMatchSnapshot();

        component.unmount();
    });

    it('should open and close properly', () => {
        //renders initial dropdown properly
        const component = mount(
            <DropDown
                value={sampleOptions[0]}
                options={sampleOptions}
            />
        );


        expect(component.state('showList')).toBeFalsy();
        expect(component.exists('.dropdown_list')).toBeFalsy();
        expect(component).toMatchSnapshot();


        //make sure dropdown opens properly
        component.find('.value_container').simulate('click');

        expect(component.state('showList')).toBeTruthy();
        expect(component.exists('.dropdown_list')).toBeTruthy();
        expect(component).toMatchSnapshot();

        //stays open
        component.find('.value_container').simulate('click');

        expect(component.state('showList')).toBeTruthy();
        expect(component.exists('.dropdown_list')).toBeTruthy();

        //make sure dropdown closes properly
        component.find('.overlay').simulate('click');

        expect(component.state('showList')).toBeFalsy();
        expect(component.exists('.dropdown_list')).toBeFalsy();
        expect(component).toMatchSnapshot();

        component.unmount();
    });

    it('should call a handler function and update in value accordinglyDictionary', () => {

        const onChangeHandler = jest.fn(value => value);

        const component = mount(
            <DropDown
                value={sampleOptions[0]}
                options={sampleOptions}
                onChange={onChangeHandler}
            />
        );

        expect(component.state('showList')).toBeFalsy();
        expect(component.exists('.dropdown_list')).toBeFalsy();

        component.find('.value_container').simulate('click');

        expect(component.state('showList')).toBeTruthy();
        expect(component.exists('.dropdown_list')).toBeTruthy();

        component.find('.dropdown_list ul').childAt(1).simulate('click');

        expect(component.state('showList')).toBeFalsy();
        expect(component.exists('.dropdown_list')).toBeFalsy();


        expect(onChangeHandler.mock.calls.length).toBe(1);
        expect(onChangeHandler.mock.results[0].value).toBe(sampleOptions[1]);

        expect(component).toMatchSnapshot();

        component.unmount();

    });
});