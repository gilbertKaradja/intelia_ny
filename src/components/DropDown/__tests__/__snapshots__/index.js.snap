// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown component should call a handler function and update in value accordinglyDictionary 1`] = `
<DropDown
  onChange={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "label": "Stop Sign",
            "value": "9",
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": Object {
            "label": "Stop Sign",
            "value": "9",
          },
        },
      ],
    }
  }
  options={
    Array [
      Object {
        "label": "Train",
        "value": "1",
      },
      Object {
        "label": "Stop Sign",
        "value": "9",
      },
      Object {
        "label": "Couch",
        "value": "joke",
      },
      Object {
        "label": "Random Sentence",
        "value": "evo993",
      },
      Object {
        "label": "Something",
        "value": "chihuahua",
      },
    ]
  }
  value={
    Object {
      "label": "Train",
      "value": "1",
    }
  }
>
  <div
    className="DropDown"
  >
    <div
      className="value_container"
      onClick={[Function]}
    >
      <div
        className="text"
      >
        Train
      </div>
      <div
        className="icon"
      >
        <img
          alt="caret"
          src="caret-down.svg"
        />
      </div>
    </div>
  </div>
</DropDown>
`;

exports[`DropDown component should open and close properly 1`] = `
<DropDown
  options={
    Array [
      Object {
        "label": "Train",
        "value": "1",
      },
      Object {
        "label": "Stop Sign",
        "value": "9",
      },
      Object {
        "label": "Couch",
        "value": "joke",
      },
      Object {
        "label": "Random Sentence",
        "value": "evo993",
      },
      Object {
        "label": "Something",
        "value": "chihuahua",
      },
    ]
  }
  value={
    Object {
      "label": "Train",
      "value": "1",
    }
  }
>
  <div
    className="DropDown"
  >
    <div
      className="value_container"
      onClick={[Function]}
    >
      <div
        className="text"
      >
        Train
      </div>
      <div
        className="icon"
      >
        <img
          alt="caret"
          src="caret-down.svg"
        />
      </div>
    </div>
  </div>
</DropDown>
`;

exports[`DropDown component should open and close properly 2`] = `
<DropDown
  options={
    Array [
      Object {
        "label": "Train",
        "value": "1",
      },
      Object {
        "label": "Stop Sign",
        "value": "9",
      },
      Object {
        "label": "Couch",
        "value": "joke",
      },
      Object {
        "label": "Random Sentence",
        "value": "evo993",
      },
      Object {
        "label": "Something",
        "value": "chihuahua",
      },
    ]
  }
  value={
    Object {
      "label": "Train",
      "value": "1",
    }
  }
>
  <div
    className="DropDown"
  >
    <div
      className="overlay"
      onClick={[Function]}
    />
    <div
      className="value_container"
      onClick={[Function]}
    >
      <div
        className="text"
      >
        Train
      </div>
      <div
        className="icon"
      >
        <img
          alt="caret"
          src="caret-down.svg"
        />
      </div>
    </div>
    <div
      className="dropdown_list"
    >
      <ul>
        <li
          className="active"
          key="1"
          onClick={[Function]}
        >
          Train
        </li>
        <li
          className=""
          key="9"
          onClick={[Function]}
        >
          Stop Sign
        </li>
        <li
          className=""
          key="joke"
          onClick={[Function]}
        >
          Couch
        </li>
        <li
          className=""
          key="evo993"
          onClick={[Function]}
        >
          Random Sentence
        </li>
        <li
          className=""
          key="chihuahua"
          onClick={[Function]}
        >
          Something
        </li>
      </ul>
    </div>
  </div>
</DropDown>
`;

exports[`DropDown component should open and close properly 3`] = `
<DropDown
  options={
    Array [
      Object {
        "label": "Train",
        "value": "1",
      },
      Object {
        "label": "Stop Sign",
        "value": "9",
      },
      Object {
        "label": "Couch",
        "value": "joke",
      },
      Object {
        "label": "Random Sentence",
        "value": "evo993",
      },
      Object {
        "label": "Something",
        "value": "chihuahua",
      },
    ]
  }
  value={
    Object {
      "label": "Train",
      "value": "1",
    }
  }
>
  <div
    className="DropDown"
  >
    <div
      className="value_container"
      onClick={[Function]}
    >
      <div
        className="text"
      >
        Train
      </div>
      <div
        className="icon"
      >
        <img
          alt="caret"
          src="caret-down.svg"
        />
      </div>
    </div>
  </div>
</DropDown>
`;

exports[`DropDown component should render without issues 1`] = `
<div
  className="DropDown"
>
  <div
    className="value_container"
    onClick={[Function]}
  >
    <div
      className="text"
    >
      Train
    </div>
    <div
      className="icon"
    >
      <img
        alt="caret"
        src="caret-down.svg"
      />
    </div>
  </div>
</div>
`;
